crates/common/logged_command/src/logged_command.rs

```diff
@@ -16,10 +19,70 @@ pub struct LoggingChild {
 
 impl LoggingChild {
     pub async fn wait_with_output(
-        self,
+        mut self,
         logger: &mut BufWriter<File>,
+        graceful_timeout: Option<Duration>,
+        forceful_timeout: Option<Duration>,
     ) -> Result<Output, std::io::Error> {
+        match graceful_timeout {
+            Some(timeout) => {
+                tokio::select! {
+                    _ = self.inner_child.wait() => {
+                        self.get_outcome(logger).await
+                    }
+                    _ = tokio::time::sleep(timeout) => {
+                        self.timeout_operation(logger, timeout, forceful_timeout).await
+                    }
+                }
+            }
+            None => self.get_outcome(logger).await,
+        }
+    }
+
+    async fn get_outcome(self, logger: &mut BufWriter<File>) -> Result<Output, std::io::Error> {
         let outcome = self.inner_child.wait_with_output().await;
+        if let Err(err) = LoggedCommand::log_outcome(&self.command_line, &outcome, logger).await {
+            error!("Fail to log the command execution: {}", err);
+        }
+        outcome
+    }
+
+    async fn timeout_operation(
+        mut self,
+        logger: &mut BufWriter<File>,
+        timeout: Duration,
+        forceful_timeout: Option<Duration>,
+    ) -> Result<Output, std::io::Error> {
+        // stop the child process by sending sigterm
+        send_sig_term(&self.inner_child).await?;
+
+        // wait to gracefully stop, if not stopped then send sigkill
+        if let Some(timeout) = forceful_timeout {
+            tokio::select! {
+                _ = self.inner_child.wait() => {
+                }
+                _ = tokio::time::sleep(timeout) => {
+                    self.inner_child.kill().await?;
+                }
+            }
+        }
+
+        let mut outcome = self.inner_child.wait_with_output().await;
+        let outcome = match outcome.as_mut() {
+            Ok(v) => {
+                v.stderr.append(
+                    &mut format!(
+                        "operation failed due to timeout: duration={}s",
+                        timeout.as_secs()
+                    )
+                    .as_bytes()
+                    .to_vec(),
+                );
+                Ok(v.clone())
```

Comment:
Why this `clone()` ? It's surprising to clone and drop `v`.

Reply:
It was under `mut ref`. To return I had to do that, but I fixed it with the correct solution `to_owned`.